src/http/SOCKSProxy.zig

```diff
@@ -0,0 +1,240 @@
+const SOCKSProxy = @This();
+const RefCount = bun.ptr.RefCount(@This(), "ref_count", SOCKSProxy.deinit, .{});
+pub const ref = SOCKSProxy.RefCount.ref;
+pub const deref = SOCKSProxy.RefCount.deref;
+
+state: SOCKSState = .init,
+destination_host: []const u8 = "",
+destination_port: u16 = 0,
+proxy_url: URL,
+allocator: std.mem.Allocator,
+ref_count: RefCount,
+
+const SOCKSState = enum {
+    init,
+    auth_handshake,
+    auth_complete,
+    connect_request,
+    connected,
+    failed,
+};
+
+const SOCKSVersion = enum(u8) {
+    v5 = 0x05,
+};
+
+const SOCKSAuthMethod = enum(u8) {
+    no_auth = 0x00,
+    gssapi = 0x01,
+    username_password = 0x02,
+    no_acceptable = 0xFF,
+};
+
+const SOCKSCommand = enum(u8) {
+    connect = 0x01,
+    bind = 0x02,
+    udp_associate = 0x03,
+};
+
+const SOCKSAddressType = enum(u8) {
+    ipv4 = 0x01,
+    domain_name = 0x03,
+    ipv6 = 0x04,
+};
+
+const SOCKSReply = enum(u8) {
+    succeeded = 0x00,
+    general_failure = 0x01,
+    connection_not_allowed = 0x02,
+    network_unreachable = 0x03,
+    host_unreachable = 0x04,
+    connection_refused = 0x05,
+    ttl_expired = 0x06,
+    command_not_supported = 0x07,
+    address_type_not_supported = 0x08,
+};
+
+pub fn create(allocator: std.mem.Allocator, proxy_url: URL, destination_host: []const u8, destination_port: u16) !*SOCKSProxy {
+    const socks_proxy = bun.new(SOCKSProxy, .{
+        .ref_count = .init(),
+        .proxy_url = proxy_url,
+        .destination_host = destination_host,
+        .destination_port = destination_port,
+        .allocator = allocator,
+    });
+
+    return socks_proxy;
+}
+
+pub fn sendAuthHandshake(this: *SOCKSProxy, comptime is_ssl: bool, socket: NewHTTPContext(is_ssl).HTTPSocket) void {
+    // SOCKS5 authentication handshake
+    // +----+----------+----------+
+    // |VER | NMETHODS | METHODS  |
+    // +----+----------+----------+
+    // | 1  |    1     | 1 to 255 |
+    // +----+----------+----------+
+    var auth_request = [_]u8{ @intFromEnum(SOCKSVersion.v5), 1, @intFromEnum(SOCKSAuthMethod.no_auth) };
+
+    _ = socket.write(&auth_request);
```

Comment:
This is definitely wrong and also slow. We need to buffer the data incase the write is incomplete.